design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,138 @@
+This feature helps user to avoid runtime failures and also increases the performances for workload deployments.
+
+### Goals
+
+Automatically classify bare metal hosts based on user-provided rules in a way that other controllers can use, without having to evaluate all of those rules themselves.
+
+## Proposal
+
+We are creating proposal based on the discussion with metal3 community on issue #351:
+https://github.com/metal3-io/baremetal-operator/issues/351
+
+We compared introspection data of Ironic with metal3 schema and we found that default  hardware
+configuration needs to be added by introducing a new CRD `HardwareClassification`.
+
+We will write a CRD that checks inspected baremetal hosts against hardware configuration and add label as "MATCHES" for matched profile into baremetal host.
+  
+### Implementation Details/Notes/Constraints
+
+Link for Existing Metal3 Specs
+Please refer metal3 spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+* Below is sample yaml of Kind HardwareClassification.
+    
+   ```yaml
+   
+   apiVersion: metal3.io.sigs.k8s.io/v1alpha1
+   kind: HardwareClassification
+   metadata:
+     name: profile1
+   spec:
+   expectedValidationConfiguration:
+      namespace: metal3
+      CPU:
+         minimumCount: 4
+         minimumSpeed: 4.3
+      Disk:
+         minimumSizeGB: 2200
+         minimumCount: 2
+      NICS:
+         minimumcount: 4
+      RAM:
+         minimumSizeGB: 32
+   ```
+   Similarly, user can mention maximum requirements for CPU, Disk, NICS and RAM.
+
+* Write a new API as hardware-classification/api/v1alpha1 and new kind(CRD) HardwareClassification.
+    
+    - Create the Schema struct for `ExpectedHardwareConfiguration` inside `HardwareClassificationSpec`,
+    in file /api/v1alpha1/hardwareClassification_types.go.
+
+    - Implement a new function fetchHost() which will fetch all baremetal hosts from Baremetal-Operator of namespace 'metal3' provided by user in CRD_yaml. We will add a watch on the baremetalhost kind for the hardware setting changes.
+
+    - Create a classification_manager.go file. Create a function which will call appropriate filter written in classification_filter.go based on minimum or maximum requiremnets specified by user in CRD_yaml. 
+
+    - The ClassificationFilters.go file will contain implementation of classification algorithms based on requirements given by user (i.e minimum or maximum). Classification will be done for minimum >= and for maximum <=  against hosts.
+
+    - Once the classification is completed, will add label if user has provided any, otherwise it will be added as "hardwareclassification.metal3.io/<profile-name>: matches". 
+
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+The following diagram shows the implementation details of Hardware Classification Controller.
+
+![Hardware Classification Controller Implementation Details](hwcc_sequence_diagram.png)
```

Comment:
The diagram shows the operation to fetch the list of hosts communicating with the baremetal-operator. That call will be a regular API call, and won't need the operator to be involved.

Reply:
okay. Yeah.